doc/xboxdrv-daemon.1
'\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH "xboxdrv-daemon " 1 2010-05-13 0.4.14 "User Commands"
.SH NAME
xboxdrv-daemon
\- A simple daemon that automatically launches xboxdrv when it detects a controller. 
.SH SYNOPSIS
'nh
.fi
.ad l
\fBxboxdrv\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[OPTION]\&... [--] {XBOXDRV} [ARGS]\&...
'in \n(.iu-\nxu
.ad b
'hy
.SH DESCRIPTION
.TP 
\*(T<\fB\-h\fR\*(T>, \*(T<\fB\-\-help\fR\*(T>
Display help text and exit.
.TP 
\*(T<\fB\-\-version\fR\*(T>
Print the version number and exit.
.TP 
\*(T<\fB\-v\fR\*(T>, \*(T<\fB\-\-verbose\fR\*(T>
Be verbose with printed output.
.TP 
\*(T<\fB\-a FILENAME, \-\-attach=FILENAME\fR\*(T>
Launch EXE when a new controller is connected
.TP 
\*(T<\fB\-d FILENAME, \-\-detach=FILENAME\fR\*(T>
Launch EXE when a controller is detached
.TP 
\*(T<\fB\-x FILENAME, \-\-xboxdrv=FILENAME\fR\*(T>
Set the location of the xboxdrv executable
.SH "RUNNING XBOXDRV VIA XBOXDRV-DAEMON"
\fBxboxdrv-daemon\fR allows you to run xboxdrv
automatically whenever you plug in a gamepad. Since
xboxdrv-daemon.py needs to know where xboxdrv is located you need
to supply the location of the xboxdrv binary manually if its not
located in \fBPATH\fR:
.PP
.nf
\*(T<% xboxdrv\-daemon \-\-xboxdrv ./xboxdrv\*(T>
.fi
.PP
You can pass arguments to xboxdrv after a double dash '--':
.PP
.nf
\*(T<% xboxdrv\-daemon \-\- \-\-buttonmap B=A,X=A,Y=A\*(T>
.fi
.PP
You can also run scripts that get started after xboxdrv is started
via:
.PP
.nf
\*(T<% xboxdrv\-daemon \e
  \-\-attach /home/juser/xboxdrv_attach \e
  \-\-detach /home/juser/xboxdrv_deatach\*(T>
.fi
.PP
This is useful in Kiosk settings, such as a media center PC, where you
might want to restart the media center to have the controller working
properly after it got attached.
.PP
Note that you can't dynamically reconfigurabilty xboxdrv
when you launch it automatically via the daemon, so using this
daemon isn't generally recommend.
.PP
When you want configurability and automatic launching, it is recommended
that you write little startup scripts for your games, such as this:
.PP
.nf
\*(T<#!/bin/sh

# Start xboxdrv and remember its PID in the variable XBOXPID
xboxdrv \-\-trigger\-as\-button \-s &amp;
XBOXPID=$!

# Give xboxdrv a second to startup and create the device
sleep 1

# Launch your favorite game
your_favorite_game

# Kill xboxdrv and wait for it to finish
kill $XBOXPID
wait $XBOXPID

# EOF #\*(T>
.fi
.PP
That way you can individually configure every game and still not have
to worry about launching xboxdrv manually.
.SH "SEE ALSO"
xboxdrv(1)


NEWS
xboxdrv 0.8.8 - (09/11/2015)
============================

* use session dbus even for user root, except when DISPLAY is not set
* fixed issue in to_float() triggering assert()
* fixed vendor/product not getting set for virtual device in daemon mode
* fixed \r\n newlines causing errors in INIParser, ignore \r like whitespace
* added device support for Rock Candy Gamepad Wired Controller, Mad
  Catz Street Fighter IV SE Fighting Stick, MadCatz GamePad, Hori
  Fighting Edge, Elcom JC-U3613M, Mad Catz Fightpad, Mad Cats Ghost
  Recon FS GamePad, Mad Catz SFxT Fightstick Pro, Razer Sabertooth,
  FUS1ON Tournament Controller, PowerA Pro Ex and others


xboxdrv 0.8.7 - (28/10/2015)
============================

* fixed USB cleanup issue on exit


xboxdrv 0.8.6 - (25/10/2015)
============================

* Buttonevent exec: double fork to avoid zombies
* fixed axis handling in FourWayRestrictorModifier
* fixed device disconnect issues
* fixed segfault due to incorrect shutdown of USB subsystem
* fixed some compiler warnings/errors
* ignore Play'n Charge kit instead of error out
* added device support for Thrustmaster Gamepad GP XID, Razer
  Sabertooth, Gamestop BB-070, Afterglow AX.1, MLG Pro Circuit
  Controller (Xbox), Rock Candy, Saitek p3600, Hori GEM Xbox
  controller


xboxdrv 0.8.5 - (07/Apr/2013)
=============================

* fixed errors on shutdown related to libusbx
* updated button mapping for --mimic-xpad-wireless and --mimic-xpad
* workaround for evdev values that don't fall into the given min/max
* added some support for Playstation 3 LED and rumble
* added support for Batarang Xbox 360 controller
* added support for Elite Glow
* added support for Hori Real Arcade Pro VX-SA
* added support for Hori SOULCALIBUR V Stick
* added support for Hori XBOX 360 EX 2 with Turbo
* added support for Logic3 Controller
* added support for Logitech F310
* added support for Logitech F510
* added support for Logitech F710
* added support for Razer Onza Tournament Edition
* added support for Street Fighter IV FightStick TE
* added support for Thrustmaster, Inc. GPX Controller
* added support for Xbox Airflo wired controller


xboxdrv 0.8.4 - (24/Jan/2012)
=============================

* fixed readded lost man-page
* added documentation for --dbus and --ff-device


xboxdrv 0.8.3 - (23/Jan/2012)
=============================

* added --dbus option to switch between system and session bus
* added sequence-key, same as cycle-key but doesn't wrap around
* added workaround for libudev older then v150
* added --ff-device DEV to select force-feedback device
* fixed force feedback getting on the wrong evdev
* fixed Xbox360 wireless LED not getting properly set on late plugin


xboxdrv 0.8.2 - (07/Aug/2011)
=============================

* added click-press, click-release, click-both button filter
* added delay button filter
* added simple Python script for response curve generation
* added PDF Afterglow AX.1 Xbox360 controller support
* fixed swapped keys in KeyAxisEventHandler


xboxdrv 0.8.1 - (20/Jul/2011)
=============================

* added const axis and button filter
* added 'generic-usb' controller type for debugging
* added cycle-key support to cycle through a sequence of buttons
* added proper merging of multiple output events to the same target
* added support for Street Fighter IV fightpad
* added support for Tron Xbox 360 controller
* added support for an unlimited number shift buttons
* fixed axis-shift getting ignored when another button is pressed
* fixed keys not getting pressed in --ui-axismap
* fixed rumble messages getting delayed
* reenabled --priority realtime


xboxdrv 0.8.0 - (26/May/2011)
=============================

* added --no-dbus option
* added ABS/axis support to macros
* added Street Fighter Fightpad support
* added example config for Microsoft Ergonomic 4000 zoom wheel
* added example config for Star Wars: Tie Fighter
* added rel-repeat value of -1 for sending an event only once
* added support for chatpad on bcdVersion 1.14 controller
* added support for relative filenames for macros
* fixed daemon mode
* fixed deadzone filter issues with MIN:MAX:SMOOTH
* fixed endian issues in Xbox360 and Xbox360 wireless controller
* fixed wireless controller not getting reset to neutral position on disconnected
* changed the meaning of --timeout 0
* removed boost::thread dependency
* switched from threads to asynchronous USB handling and glib


xboxdrv 0.7.3 - (16/Mar/2011)
=============================

* fixed installation of xboxdrvctl
* added special REL repeat value '-1' to avoid jaggy mouse emulation movement
* added "rel-repeat:" axis-event for better scroll wheel emulation
* added --no-dbus, to disable D-Bus support
* added xbmc.xboxdrv, a config fine tuned for XBMC
* added zsnes.xboxdrv, a config file for ZSNES
* reenabled USBReadThread to work around ignored input events
* '--daemon' is broken in this release


xboxdrv 0.7.2 - (06/Mar/2011)
=============================

* added --mimic-xpad-wireless
* added --device-usbid and --device-usbids
* added D-Bus support to the daemon
* added xboxdrvctl to control the daemon
* added better handling of rounding errors when sending REL_
* added dpad-restrictor for X, Y or fourway restriction on the dpad
* added statistic modifier that counts how often buttons are pressed
* added support for multiple wireless controller in daemon
* added --priority to enable realtime scheduling priority
* added custom configuration for XBMC, see examples/xbmc.xboxdrv
* changed default timeout to 10msec, for more responsive mouse emulation
* changed meaning of -m, its now a shortcut for --modifier, not --mouse
* improved INIParser, can now handle KEY_#NUM style names without quotes
* improved mouse emulation
* removed obsolete xboxdrv-daemon script
* renamed --device-name to --device-names
* renamed --name to --device-name


xboxdrv 0.7.1 - (30/Jan/2011)
=============================

* added --match-group

tools/xboxdrv-daemon.py
#!/usr/bin/env python
##
##  Xbox360 USB Gamepad Userspace Driver
##  Copyright (C) 2009 Ingo Ruhnke <grumbel@gmx.de>
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see <http://www.gnu.org/licenses/>.

import string
import time
import dbus
from dbus.mainloop.glib import DBusGMainLoop
import gobject
import sys
import os
import signal
import subprocess
from optparse import OptionParser

class DeviceManager:
    def __init__(self, xboxdrv_device_list=[], xboxdrv="xboxdrv", attach=None, detach=None, xboxdrv_args=[], verbose = False):
        self.xboxdrv_device_list = xboxdrv_device_list
        self.xboxdrv_bin   = xboxdrv
        self.attach_script = attach
        self.detach_script = detach
        self.xboxdrv_args  = xboxdrv_args
        self.verbose = verbose

        self.processes = {}
        self.bus = dbus.SystemBus()

        self.bus.add_signal_receiver(self.device_added,
                                     'DeviceAdded',
                                     'org.freedesktop.Hal.Manager',
                                     'org.freedesktop.Hal',
                                     '/org/freedesktop/Hal/Manager')
        
        self.bus.add_signal_receiver(self.device_removed,
                                     'DeviceRemoved',
                                     'org.freedesktop.Hal.Manager',
                                     'org.freedesktop.Hal',
                                     '/org/freedesktop/Hal/Manager')

        hal_manager_obj = self.bus.get_object("org.freedesktop.Hal", "/org/freedesktop/Hal/Manager")
        hal_manager     = dbus.Interface(hal_manager_obj, "org.freedesktop.Hal.Manager")

        if self.verbose:
            print "xboxdrv-daemon: searching for connected controllers"

        # find xboxdrv devices that are already connected
        for udi in hal_manager.FindDeviceStringMatch('info.subsystem', 'usb_device'):
            self.device_added(udi)

    def udi_to_device(self, udi):
        # uid: /org/freedesktop/Hal/devices/usb_device_45e_28e_13FEF2D
        return self.bus.get_object("org.freedesktop.Hal", udi)
    
    def device_added(self, udi):
        # print "xboxdrv-daemon: usb device connected: %s" % udi

        device = self.udi_to_device(udi)
        device_if = dbus.Interface(device, 'org.freedesktop.Hal.Device')

        if device_if.GetPropertyString('info.subsystem') == 'usb_device':
            vendor_id  = device_if.GetPropertyInteger('usb_device.vendor_id')
            product_id = device_if.GetPropertyInteger('usb_device.product_id')

            device_type = self.is_xboxdrv_device(vendor_id, product_id)

            if device_type:
                if self.verbose:
                    print "xboxdrv-daemon: controller connected: %s" % udi

                bus_number = device_if.GetPropertyInteger('usb_device.bus_number')
                dev_number = device_if.GetPropertyInteger('usb_device.linux.device_number')
                self.xboxdrv_launch(udi, bus_number, dev_number, device_type[0])
            
    def device_removed(self, udi):
        # print "xboxdrv-daemon: usb device disconnected: %s" % udi
        if self.verbose and self.processes.has_key(udi):
            print "xboxdrv-daemon: controller disconnected: %s" % udi
        self.xboxdrv_kill(udi)

    def xboxdrv_launch(self, udi, bus, dev, type):
        args = [self.xboxdrv_bin,
                '--silent', 
                '--quiet',
                '--device-by-path', "%03d:%03d" % (bus, dev), 
                '--type', type] + self.xboxdrv_args

        if self.verbose:
            print "xboxdrv-daemon: launching: %s" % string.join(args)

        # self.processes[udi] = os.spawnvp(os.P_NOWAIT, self.xboxdrv_bin, args)
        try: 
            self.processes[udi] = subprocess.Popen(args).pid
        except OSError, err:
            raise Exception("xboxdrv-daemon: couldn't launch '%s', set the location of xboxdrv with '-x'" % self.xboxdrv_bin)

        if self.verbose:
            print "xboxdrv-daemon: launched: pid: %d - %s" % (self.processes[udi], udi)

        if self.attach_script:
            time.sleep(2) # give xboxdrv time to start up
            if self.verbose:
                print "xboxdrv-daemon: executing attach script: %s" % self.attach_script
            os.system(self.attach_script)

    def xboxdrv_kill(self, udi):
        if self.processes.has_key(udi):
            if self.verbose:
                print "xboxdrv-daemon: killing: pid: %d - %s" % (self.processes[udi], udi)

            pid = self.processes[udi]
            os.kill(pid, signal.SIGINT)
            os.waitpid(pid, 0)
            del self.processes[udi]

            if self.detach_script:
                if self.verbose:
                    print "xboxdrv-daemon: executing detach script: %s" % self.detach_script
                os.system(self.detach_script)

    def shutdown(self):
        for (udi, pid) in self.processes.items():
            self.xboxdrv_kill(udi)
            
    def is_xboxdrv_device(self, arg_vendor_id, arg_product_id):
        for (type, vendor_id, product_id, name) in self.xboxdrv_device_list:
            # print arg_vendor_id,  vendor_id, arg_product_id, product_id
            if arg_vendor_id == vendor_id and arg_product_id == product_id:
                return (type, vendor_id, product_id, name)
        return None
    
if __name__ == '__main__':
    # parse options
    parser = OptionParser(usage = "%prog [OPTIONS] -- [XBOXDRV ARGS]",
                          version = "0.1",
                          description = "A simple daemon that automatically launches xboxdrv.")
    parser.add_option('-a', '--attach', metavar='FILENAME',
                      help="Launch EXE when a new controller is connected")
    parser.add_option('-d', '--detach', metavar='FILENAME',
                      help="Launch EXE when a controller is detached")
    parser.add_option('-x', '--xboxdrv', metavar='FILENAME', default="xboxdrv",
                      help="Set the location of the xboxdrv executable")
    parser.add_option('-v', '--verbose', action='store_true', default=False,
                      help="Be verbose with printed output")
    (opts, args) = parser.parse_args()

    # Read list of supported devices from xboxdrv
    try:
        xboxdrv_device_list = []
        for i in subprocess.Popen([opts.xboxdrv, '--list-supported-devices'], 
                                  stdout=subprocess.PIPE).stdout:
            lst = i.rstrip().split(' ', 3)
            xboxdrv_device_list.append((lst[0],
                                        int(lst[1], 16),
                                        int(lst[2], 16),
                                        lst[3]))
    except OSError, err:
        raise Exception("xboxdrv-daemon: couldn't launch '%s', set the location of xboxdrv with '-x'" % opts.xboxdrv)

    DBusGMainLoop(set_as_default=True)

    mgr  = DeviceManager(xboxdrv_device_list = xboxdrv_device_list,
                         xboxdrv = opts.xboxdrv, xboxdrv_args = args, 
                         attach = opts.attach, detach = opts.detach,
                         verbose = opts.verbose)
    loop = gobject.MainLoop()

    try:
        loop.run()

    except KeyboardInterrupt:
        if opts.verbose:
            print "xboxdrv-daemon: keyboad interrupt received, shutting down"

        loop.quit()
        mgr.shutdown()
        sys.exit(0)

# EOF #
